ui/vr_manager.js

```diff
@@ -0,0 +1,496 @@
+      // See: https://dev.to/li/how-to-requestpermission-for-devicemotion-and-deviceorientation-events-in-ios-13-46g2
+      if (typeof DeviceMotionEvent.requestPermission == 'function') {
+        const userGestureListener = () => {
+          DeviceMotionEvent.requestPermission().then((newPermissionState) => {
+            if (newPermissionState !== 'granted' ||
+                deviceOrientationListener) {
+              return;
+            }
+            deviceOrientationListener = true;
+            this.setupDeviceOrientationListener_();
+          });
+        };
+        DeviceMotionEvent.requestPermission().then((permissionState) => {
+          this.eventManager_.unlisten(
+              this.container_, 'click', userGestureListener);
+          this.eventManager_.unlisten(
+              this.container_, 'mouseup', userGestureListener);
+          if (navigator.maxTouchPoints > 0) {
+            this.eventManager_.unlisten(
+                this.container_, 'touchend', userGestureListener);
+          }
+          if (permissionState !== 'granted') {
+            this.eventManager_.listenOnce(
+                this.container_, 'click', userGestureListener);
+            this.eventManager_.listenOnce(
+                this.container_, 'mouseup', userGestureListener);
+            if (navigator.maxTouchPoints > 0) {
+              this.eventManager_.listenOnce(
+                  this.container_, 'touchend', userGestureListener);
+            }
+            return;
+          }
+          deviceOrientationListener = true;
+          this.setupDeviceOrientationListener_();
+        }).catch(() => {
+          this.eventManager_.unlisten(
+              this.container_, 'click', userGestureListener);
+          this.eventManager_.unlisten(
+              this.container_, 'mouseup', userGestureListener);
+          if (navigator.maxTouchPoints > 0) {
+            this.eventManager_.unlisten(
+                this.container_, 'touchend', userGestureListener);
+          }
+          this.eventManager_.listenOnce(
+              this.container_, 'click', userGestureListener);
+          this.eventManager_.listenOnce(
+              this.container_, 'mouseup', userGestureListener);
+          if (navigator.maxTouchPoints > 0) {
+            this.eventManager_.listenOnce(
+                this.container_, 'touchend', userGestureListener);
+          }
+        });
+      } else {
+        deviceOrientationListener = true;
+        this.setupDeviceOrientationListener_();
+      }
+    }
+  }
+
+  /**
+   * @private
+   */
+  setupDeviceOrientationListener_() {
+    this.eventManager_.listen(window, 'deviceorientation', (e) => {
+      const event = /** @type {!DeviceOrientationEvent} */(e);
+      let alphaDif = (event.alpha || 0) - this.prevAlpha_;
+      let betaDif = (event.beta || 0) - this.prevBeta_;
+      let gammaDif = (event.gamma || 0) - this.prevGamma_;
+
+      if (Math.abs(alphaDif) > 10) {
```

Comment:
This is so if you make a very fast movement of your device, it won't rotate at all, right?

I wonder if it would feel better to set all of the dif values to 0 if any one of them is too high.

As-is, if you're rotating your device very quickly in one axis but slowly in the other, the viewport will rotate in the direction of the slow movement but not the fast movement. That might feel weird, I'm not sure.

Reply:
I have changed it so that there is no movement in that case.